pkg/envtest/apiserver.go

```diff
@@ -0,0 +1,443 @@
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	settingsv1alpha1 "github.com/gardener/gardener/pkg/apis/settings/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+	"github.com/gardener/gardener/pkg/utils/retry"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+)
+
+const (
+	envGardenerAPIServerBin = "TEST_ASSET_GARDENER_APISERVER"
+	waitPollInterval        = 100 * time.Millisecond
+)
+
+// GardenerAPIServer knows how to start, register and stop a temporary gardener-apiserver instance.
+type GardenerAPIServer struct {
+	// EtcdURL is the etcd URL that the APIServer should connect to (defaults to the URL of the envtest etcd).
+	EtcdURL *url.URL
+	// CertDir is a path to a directory containing whatever certificates the APIServer needs.
+	// If left unspecified, then Start will create a temporary directory and generate the needed
+	// certs and Stop will clean it up.
+	CertDir string
+	// caCert is the certificate of the CA that signed the GardenerAPIServer's serving cert.
+	caCert *secrets.Certificate
+	// restConfig is used to setup and register the APIServer with the envtest kube-apiserver.
+	restConfig *rest.Config
+	// Path is the path to the gardener-apiserver binary, can be set via TEST_ASSET_GARDENER_APISERVER.
+	// If Path is unset, gardener-apiserver will be started in-process.
+	Path string
+	// SecurePort is the secure port that the APIServer should listen on.
+	// If this is not specified, we default to a random free port on localhost.
+	SecurePort int
+	// listenURL is the URL we end up listening on.
+	listenURL *url.URL
+	// Args is a list of arguments which will passed to the APIServer binary.
+	// If not specified, the minimal set of arguments to run the APIServer will
+	// be used.
+	Args []string
+	// StartTimeout, StopTimeout specify the time the APIServer is allowed to
+	// take when starting and stoppping before an error is emitted.
+	// If not specified, these default to 20 seconds.
+	StartTimeout time.Duration
+	StopTimeout  time.Duration
+	// Out, Err specify where APIServer should write its StdOut, StdErr to.
+	// If not specified, the output will be discarded.
+	Out io.Writer
+	Err io.Writer
+	// HealthCheckEndpoint is the path of the healthcheck endpoint (defaults to "/healthz").
+	// It will be polled until receiving http.StatusOK (or StartTimeout occurs), before
+	// returning from Start.
+	HealthCheckEndpoint string
+
+	// terminateFunc holds a func that will terminate this GardenerAPIServer.
+	terminateFunc func()
+	// exited is a channel that will be closed, when this GardenerAPIServer exits.
+	exited chan struct{}
+}
+
+// Start brings up the GardenerAPIServer, waits for it to be healthy and registers Gardener's APIs.
+func (a *GardenerAPIServer) Start() error {
```

Comment:
Cosmetics: Shouldn't this rather be

```suggestion
func (g *GardenerAPIServer) Start() error {
```